src/persist-client/src/batch.rs

```diff
@@ -1338,48 +1338,84 @@ pub(crate) fn validate_truncate_batch<T: Timestamp>(
     Ok(())
 }
 
-#[derive(Debug, Default)]
-pub(crate) struct PartDeletes(BTreeSet<PartialBatchKey>);
+#[derive(Debug)]
+pub(crate) struct PartDeletes<T> {
+    hollow_parts: BTreeSet<PartialBatchKey>,
+    hollow_runs: BTreeMap<PartialBatchKey, HollowRunRef<T>>,
+}
+
+impl<T> Default for PartDeletes<T> {
+    fn default() -> Self {
+        Self {
+            hollow_parts: Default::default(),
+            hollow_runs: Default::default(),
+        }
+    }
+}
 
-impl PartDeletes {
+impl<T: Timestamp> PartDeletes<T> {
     // Adds the part to the set to be deleted and returns true if it was newly
     // inserted.
-    pub fn add<T>(&mut self, part: &RunPart<T>) -> bool {
+    pub fn add(&mut self, part: &RunPart<T>) -> bool {
         match part {
-            RunPart::Many(_) => {
-                todo!("walk the tree of parts and delete them individually.")
-            }
-            RunPart::Single(BatchPart::Hollow(x)) => self.0.insert(x.key.clone()),
+            RunPart::Many(r) => self.hollow_runs.insert(r.key.clone(), r.clone()).is_none(),
+            RunPart::Single(BatchPart::Hollow(x)) => self.hollow_parts.insert(x.key.clone()),
             RunPart::Single(BatchPart::Inline { .. }) => {
                 // Nothing to delete.
                 true
             }
         }
     }
 
-    pub async fn delete(
-        self,
-        blob: &Arc<dyn Blob>,
-        shard_id: ShardId,
-        metrics: &Arc<RetryMetrics>,
-    ) {
-        let deletes = FuturesUnordered::new();
-        for key in self.0 {
-            let metrics = Arc::clone(metrics);
-            let blob = Arc::clone(blob);
-            deletes.push(async move {
-                let key = key.complete(&shard_id);
-                retry_external(&metrics, || blob.delete(&key)).await;
-            });
+    pub fn contains(&self, part: &RunPart<T>) -> bool {
+        match part {
+            RunPart::Many(r) => self.hollow_runs.contains_key(&r.key),
+            RunPart::Single(BatchPart::Hollow(x)) => self.hollow_parts.contains(&x.key),
+            RunPart::Single(BatchPart::Inline { .. }) => false,
         }
-        let () = deletes.collect().await;
     }
-}
 
-impl Deref for PartDeletes {
-    type Target = BTreeSet<PartialBatchKey>;
-    fn deref(&self) -> &Self::Target {
-        &self.0
+    pub fn is_empty(&self) -> bool {
+        self.hollow_runs.is_empty() && self.hollow_parts.is_empty()
```

Comment:
nit: Maybe deconstruct `Self{hollow_runs, hollow_parts} = self` to be defensive against another field being added later?